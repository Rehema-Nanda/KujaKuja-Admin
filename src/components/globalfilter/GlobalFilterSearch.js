import React from "react";

import { Input } from "reactstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import PropTypes from "prop-types";

import { debounce } from "lodash";

export default class GlobalFilterSearch extends React.Component {
    handleSearchDebounced = debounce((value) => {
        const { setData, toggleRefresh } = this.props;

        setData("searchString", value);
        toggleRefresh();
    }, 500);

    constructor(props) {
        super(props);
        this.onChange = this.onChange.bind(this);
    }

    componentDidMount() {
        const { clearSearch } = this.props;

        if (clearSearch) {
            this.clearSearchValue();
        }
    }

    componentDidUpdate(prevProps) {
        const { clearSearch } = this.props;

        if (prevProps.clearSearch !== clearSearch) {
            this.clearSearchValue();
        }
    }

    componentWillUnmount() {
        this.handleSearchDebounced.cancel();
    }

    onChange(event) {
        // ref: https://stackoverflow.com/questions/23123138/perform-debounce-in-react-js
        // note that we are NOT passing the event object or using event.persist()
        // https://stackoverflow.com/questions/23123138/perform-debounce-in-react-js#comment83607053_24679479
        this.handleSearchDebounced(event.target.value);
    }

    clearSearchValue = () => {
        const { setClearSearchFalse } = this.props;

        this.searchInput.value = "";
        setClearSearchFalse();
    }

    render() {
        const { placeholder, defaultSearchValue } = this.props;

        return (
            <div>
                <div className="map-controls-search-icon"><FontAwesomeIcon icon="search" /></div>
                <Input
                    type="text"
                    innerRef={e => this.searchInput = e}
                    name=""
                    onChange={this.onChange}
                    id="search"
                    className="map-controls-search"
                    placeholder={placeholder || "Search"}
                    defaultValue={defaultSearchValue}
                />
            </div>
        );
    }
}

GlobalFilterSearch.propTypes = {
    clearSearch: PropTypes.bool,
    setClearSearchFalse: PropTypes.func.isRequired,
    placeholder: PropTypes.string,
    setData: PropTypes.func.isRequired,
    toggleRefresh: PropTypes.func.isRequired,
    defaultSearchValue: PropTypes.string,
};

GlobalFilterSearch.defaultProps = {
    clearSearch: false,
    placeholder: "",
    defaultSearchValue: "",
};
