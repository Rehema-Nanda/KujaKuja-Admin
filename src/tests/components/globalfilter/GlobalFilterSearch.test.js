import React from "react";
import { shallow } from "enzyme";
import GlobalFilterSearch from "../../../components/globalfilter/GlobalFilterSearch";

describe("<GlobalFilterSearch/>", () => {
    let wrapper;
    const testProps = {
        setData: jest.fn(),
        toggleRefresh: jest.fn(),
    };

    beforeEach(() => {
        wrapper = shallow(<GlobalFilterSearch {...testProps} />);
    });

    it("renders correctly", () => {
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it("should call handleSearchDebounced function when input changes", () => {
        const testFunc = jest.spyOn(wrapper.instance(), "handleSearchDebounced");
        const event = {
            target: {
                value: "someValue",
            },
        };
        const datePicker = wrapper.find("Input");
        datePicker.simulate("change", event);
        expect(testFunc).toHaveBeenCalled();
    });
});
