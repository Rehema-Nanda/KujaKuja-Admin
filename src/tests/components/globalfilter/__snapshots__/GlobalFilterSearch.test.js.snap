// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GlobalFilterSearch/> renders correctly 1`] = `
"<div>
  <div className=\\"map-controls-search-icon\\">
    <FontAwesomeIcon icon=\\"search\\" />
  </div>
  <Input type=\\"text\\" innerRef={[Function: innerRef]} name=\\"\\" onChange={[Function: bound onChange]} id=\\"search\\" className=\\"map-controls-search\\" placeholder=\\"Search\\" defaultValue=\\"\\" />
</div>"
`;
